icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+     * the tasks are complete. This loops until a kill signal is sent via the class's destructor.
+     */
+    void run();
+
+    /**
+     * @brief Get free slot to insert new task to be executed. This isn't a blocking function - it checks all worker's 
+     * tasks and if no free one is found a nullptr is returned.
+     * @param task_slot - pointer to the internal task of the worker to allow the user to edit in the new task.
+     * @return boolean value indication of the current content of the task - true indicates that `task_slot` holds a 
+     * previous result that should be handled by the user before writing the new task.
+     */
+    bool get_free_task(Task*& task_slot);
+
+    /**
+     * @brief Get task that holds previous result to be handled by the user. This isn't a blocking function - it checks 
+     * all worker's tasks and if no completed one is found a nullptr is returned.
+     * @param completed_task - pointer to the internal task holding a previous result. 
+     * Again if using completed_task to assign additional tasks, the existing result must be handled before hand.
+     */
+    void get_completed_task(Task*& completed_task);
+
+    /**
+     * @brief Check if all worker's tasks are idle (free with no result to be handled)
+     */
+    bool are_all_idle();
+  private:
+    std::thread task_executor; // Thread to be run parallel to main
+    std::vector<Task> mTasksFifo; // vector containing the worker's task. a Vector is used to allow buffering.
+    int tail; // Tail (input) idx of the fifo above. Checks for free task start at this idx.
+    int head; // Head (output) idx of the fifo above, the thread works on task at this idx.
+    bool kill; // boolean to flag from main to the thread to finish.
+  };
+
+  std::vector<Worker> workers; // Vector of workers/thread to be ran simultaneously.
+};
+
+/**
+ * @class TaskBase
+ * @brief abstract base for the task supported by TasksManager above.
+ */
+class TaskBase {
+public:
+  /**
+   * @brief constructor fo `TaskBase`.
+   */
+  TaskBase() : status(IDLE), father_fifo_tail(nullptr) {}
+
+  /**
+   * @brief function to be executed by `TasksManager`. Implemented by child class.
+   */
+  virtual void execute() = 0;
+
+  /**
+   * @brief Signal for the `Worker` owning the task that it is ready to be executed. 
+   * Also increase's the workers tail idx.
+   * User function.
+   */
+  void dispatch();
+
+  // Getters and setter for the various states of the tasks `status`.
+  inline bool is_ready_for_work(); // DISPATCHED
+  inline bool can_push_task(); // IDLE or PENDING_RESULT
+  inline bool has_result();
+  inline bool is_idle();
+  inline void set_working();
+  inline void set_pending_result();
+  inline void set_handled_result();
+
+  /**
+   * @brief link this task with the tail idx of the Worker it belongs to. Used only once after construction.
+   */
+  inline void link_father_tail(int* tail_pointer) { father_fifo_tail = tail_pointer; } 
```

Comment:
this seems wrong to me. You should avoid this coupling with the fifo class because you couple the Task to a fifo implementation now but task doesn't need this constraint.